crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -128,37 +127,172 @@ impl<S: States> NextState<S> {
-                    to: entered.clone(),
-                })
-                .ok();
-            world.try_run_schedule(OnEnter(entered)).ok();
+/// Label to order all [`OnEnter`], [`OnTransition`], and [`OnExit`] systems.
+///
+/// All `OnEnter` will run before all `OnTransition`,
+/// and all `OnTransition` will run before all `OnExit`.
+///
+/// The order of `OnEnter` systems relative to each other (and so on) is ambiguous by default.
+#[derive(SystemSet, Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum StateTransitionSet {
+    /// Call `OnEnter` once on first run.
+    RunEnterOnce,
+    /// Change all the states before invoking transition systems here.
+    BeforeTransition,
+    /// All [`OnEnter`] schedules are executed in this set.
+    OnEnter,
+    /// All [`OnTransition`] schedules are executed in this set.
+    OnTransition,
+    /// All [`OnExit`] schedules are executed in this set.
+    OnExit,
+    /// Cleanup `PrevState` after all transition systems have run.
+    AfterTransition,
+}
+
+/// Run all state transition systems for the given state type.
+///
+/// This function is deprecated. It is not used internally by Bevy,
+/// and will be removed in a future version.
+#[deprecated(since = "0.13.0")]
+pub fn apply_state_transition<S: States>(
+    world: &mut World,
+    next_state_changed_states: &mut SystemState<(Res<State<S>>, Res<NextState<S>>)>,
+    before_transition_states: &mut SystemState<(
+        ResMut<PrevState<S>>,
+        ResMut<State<S>>,
+        ResMut<NextState<S>>,
+    )>,
+    after_transition_states: &mut SystemState<ResMut<PrevState<S>>>,
+) {
+    let state_changed = {
+        let (state, next) = next_state_changed_states.get(world);
+        next_state_changed(state, next)
+    };
+    if state_changed {
+        {
+            let (prev, state, next) = before_transition_states.get_mut(world);
+            apply_before_transition::<S>(prev, state, next);
+        }
+        run_on_enter::<S>(world);
+        run_on_transition::<S>(world);
+        run_on_exit::<S>(world);
+        {
+            let states = after_transition_states.get_mut(world);
+            apply_after_transition::<S>(states);
         }
     }
 }
+
+/// Configure systems running state transition callbacks.
+///
+/// [`OnEnter<S>`] schedule will be run once on first run.
+/// And on state change, the following schedules will be run in order:
+/// * [`OnExit<S>`]
+/// * [`OnTransition<S>`]
+/// * [`OnEnter<S>`]
+///
+/// This is called by `App::add_state`, and is not meant to be used directly.
+pub fn apply_state_transition_systems<S: States>() -> SystemConfigs {
+    // Extra `.chain()` calls case `apply_state_transition_systems` is used directly
+    // without properly configuring the `StateTransitionSet` ordering.
+    IntoSystemConfigs::into_configs(
+        (
+            run_enter_schedule::<S>
+                .run_if(common_conditions::run_once())
+                .in_set(StateTransitionSet::RunEnterOnce),
```

Comment:
I'm thinking maybe we shouldn't include the `run_enter_schedule.run_if(run_once())` system in `apply_state_transition_systems`. I would like it if you could add `apply_state_transition_systems<T>` to multiple schedules in order to allow multiple state transitions per frame, but the system that runs once kind of breaks that.

We should probably not include this system in then function, and have `App::add_state` add it separately.

Reply:
OK.